### ISSUE TITLE
Unable to install Plugins on v2

### ISSUE THREAD

**Issue by willfarrell** [geerlingguy/ansible-role-jenkins#70]
```
TASK [geerlingguy.jenkins : Install Jenkins plugins.] **************************
failed: [*******] (item=cloudbees-folder) => {"changed": true, "cmd": ["java", "-jar", "/opt/jenkins-cli.jar", "-s", "http://localhost:8080/", "install-plugin", "cloudbees-folder", "--username", "admin", "--password", "admin"], "delta": "0:00:00.151930", "end": "2016-07-25 19:42:42.309683", "failed": true, "item": "cloudbees-folder", "rc": 255, "start": "2016-07-25 19:42:42.157753", "stderr": "java.io.IOException: No X-Jenkins-CLI2-Port among [null, X-Required-Permission, X-Jenkins, X-You-Are-In-Group, X-Hudson, Content-Length, Expires, X-You-Are-Authenticated-As, X-Permission-Implied-By, Set-Cookie, Server, X-Content-Type-Options, Date, X-Jenkins-Session, Content-Type]\n\tat hudson.cli.CLI.getCliTcpPort(CLI.java:284)\n\tat hudson.cli.CLI.<init>(CLI.java:128)\n\tat hudson.cli.CLIConnectionFactory.connect(CLIConnectionFactory.java:72)\n\tat hudson.cli.CLI._main(CLI.java:473)\n\tat hudson.cli.CLI.main(CLI.java:384)\n\tSuppressed: java.io.IOException: Server returned HTTP response code: 403 for URL: http://localhost:8080/cli\n\t\tat sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1628)\n\t\tat hudson.cli.FullDuplexHttpStream.<init>(FullDuplexHttpStream.java:78)\n\t\tat hudson.cli.CLI.connectViaHttp(CLI.java:152)\n\t\tat hudson.cli.CLI.<init>(CLI.java:132)\n\t\t... 3 more", "stdout": "", "stdout_lines": [], "warnings": []}
	to retry, use: --limit @./jenkins.retry
```

Tried rerunning after creating the default user admin and same result. Guessing the hook changed in v2.

**Comment by leifmadsen** [comment]
Ansible v2? I'm installing plugins with Ansible v2 on my system and it seems to work fine...

```
jenkins_plugins:
  - build-timeout
  - scp
  - credentials-binding
  - ssh-slaves
  - xtrigger
  - copy-to-slave
```

**Comment by willfarrell** [comment]
Sorry, I ment Jenkins v2. Hooks were in ref to Jenkins, should of been more clear. I'm also using ansible v2.

**Comment by leifmadsen** [comment]
Odd... seems to work on Jenkins 2.14 for me...

**Comment by willfarrell** [comment]
Interesting. I just spun up a fresh debian 8 image at AWS and only ran the jenkins role, got the same error. 

```yml
---
- hosts: jenkins
  become: true
  vars:
    - jenkins_plugins:
      - "cloudbees-folder"
  roles:
    - geerlingguy.jenkins
```

Thanks for testing, though.

**Comment by leifmadsen** [comment]
Strange. FWIW I'm testing / deploying on CentOS.

**Comment by leifmadsen** [comment]
@willfarrell I just did a full teardown and rebuild test, and now it looks like I'm hitting the same issue. I don't think the issue is with the playbook (because I didn't change my Ansible environment, just the VM environment).

I suspect it's a Jenkins issue with 2.15, because I was running fine on 2.14. I won't be able to investigate any further as I'm on vacation starting now (for a week). Based on what I've been seeing though, I think this is a Jenkins bug, not an Ansible role issue.

**Comment by willfarrell** [comment]
Okay, thanks for looking into it. I'll post it over at jenkins. Have fun on vacation.

**Comment by leifmadsen** [comment]
@willfarrell sorry again... I looked further into this. I honestly have zero idea how this was working at all before for me. The `plugins.yml` task file has the Install Jenkins plugins. play and that uses a command that calls the jenkins-cli.jar file.

The arguments in use are `--username` and `--password`. According to this page (https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+CLI) those arguments were removed in 1.419.

Not sure if the jenkins-cli.jar lines up exactly with the same version numbers as the full release. If so, I'm super confused how I had this working at all previously.

**Comment by geerlingguy** [comment]
Version pinning will be doable in the next release... just working on a couple more tweaks before tagging a release to Galaxy. Hopefully we can get to the bottom of this!

**Comment by leifmadsen** [comment]
@geerlingguy so I know the cause and am working on a fix now. Hoepfully I'll have a PR today. I'd like to consider this a blocker for the next tagged version.

**Comment by geerlingguy** [comment]
@leifmadsen - Thanks! And I'm 50/50 on considering plugin install a blocker—there's a reason why I made it default to not installing anything, because it's been so fragile over the course of this role's lifetime (Jenkins seems to not have CLI/automated config management of itself as a high priority usually...).

But if you can get it working reliably (and want to add back a test of installing a plugin in one of the Travis environments), that'd be awesome!

**Comment by geerlingguy** [comment]
@leifmadsen - Also note that I just merged in some other changes to follow up on #75 and get tests running on CentOS 7 (CentOS 6 tests are sloooow and inexplicably fail from time to time on Travis' Docker environment).

**Comment by leifmadsen** [comment]
@geerlingguy awesome! Yea if the tests don't execute consistently, then there is kind of no point to them.

OK, after talking to @yrobla I understand now what the issue is. Basically, the version pinning introduced a regression that she had *just* fixed :) I'm now working through how best to address this. Having looked at the code some more, there are a few things I don't like about the version pinning as it stands now, and I want to address those at the same time (the use of /tmp/ to download the files, no clean up of the files, etc).

@willfarrell basically, this issue is already solved by #72 -- the issue was that the `--username` and `--password` options are not available to the `jenkins-cli` application when authentication isn't setup. Yolanda just recently added that functionality, and now it should work (but it won't work with version pinning right now -- I'm going to fix that soon).

**Comment by willfarrell** [comment]
Thanks everyone. I'm in no rush for it. I can hammer it with some test once everything is ready to go.

**Comment by leifmadsen** [comment]
OK, #77 is ready for review.

**Comment by willfarrell** [comment]
I seem to be getting the same result with your changes (#77), even with version pinning to `2.14`.

**Comment by leifmadsen** [comment]
@willfarrell is this a clean install? Try removing the Jenkins package (with yum or apt) and then running again.

**Comment by willfarrell** [comment]
@leifmadsen Yeah, I do a clean Debian ami every time.

**Comment by leifmadsen** [comment]
@willfarrell awesome thanks, I did the same thing locally (it takes a while to spin up in my environment). I just reproduced as well. I'll look into it this afternoon and hopefully have it sorted then. Thanks for the feedback!

**Comment by leifmadsen** [comment]
Well, turns out I was wrong about the problem I saw. I hadn't switched over to using my own repo (which has this changed merged down). I just tested it again, and it seems to work fine with version pinning on my setup. I am using this in my `requirements.yml`:

```
- src: https://github.com/leifmadsen/ansible-role-jenkins.git
  version: master
  name: leifmadsen.jenkins
```

The main difference I see is that I'm using CentOS and you're using Debian. Can you try with my branch on Debian with a fresh deploy, just to see if it's a Debian thing, or maybe something I did wrong in the play?

I'm also going to attach some changes here to `.travis.yml` as well as it looks like there are tests to deploy a module, but it's just not being called in the Travis-CI.

**Comment by willfarrell** [comment]
Without version pinning: same result.

With: `jenkins_version: 2.14`:
```bash
TASK [galaxy.jenkins : Install our specific version of Jenkins.] ***************
fatal: [107.22.139.154]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_Ow0AIZ/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_Ow0AIZ/ansible_module_apt.py\", line 798, in main\r\n    force=force_yes, dpkg_options=p['dpkg_options'])\r\n  File \"/tmp/ansible_Ow0AIZ/ansible_module_apt.py\", line 538, in install_deb\r\n    diff[\"prepared\"] += '\\n\\n' + out\r\nKeyError: 'prepared'\r\n", "msg": "MODULE FAILURE", "parsed": false}
	to retry, use: --limit @jenkins.retry
```

**Comment by leifmadsen** [comment]
@willfarrell honestly that is kind of looking like a bug in the Ansible module at first glance. What version of Ansible are you running with?

**Comment by willfarrell** [comment]
That's what it looks like. I was updating while you posted. `2.0.x` -> `2.1.1.0`
Same error for both use cases now. hmmmm

**Comment by leifmadsen** [comment]
Well that's interesting. Did a quick Google search, found this:

https://github.com/ansible/ansible-modules-core/issues/3752

Looks to be resolved in the code, but possibly not released yet.

**Comment by willfarrell** [comment]
lol, well that's just super. Maybe we'll have to wait this bug out.

**Comment by leifmadsen** [comment]
@willfarrell try going back to stable-2.0 as I saw someone else revert back to an older version of the 2.x deploy where it worked. Maybe it was a regression in a sub-release of 2.0. Worth a shot any ways.

I'm pretty confident based on my testing that my changes here all work fine. Seems to just be the version pinning that is broken on Debian (regardless of this change directly, since version pinning was added to this role in another commit).

**Comment by leifmadsen** [comment]
@geerlingguy, @willfarrell, I've updated the Travis-CI yaml file to perform a larger set of tests (there were some tests that weren't actually being deployed in Travis-CI but were available in the `tests/` directory).

I just realized I could probably add a version-pinning test as well. I might do that in a moment. It will probably fail on Debian though based on the feedback @willfarrell is providing :)

**Comment by leifmadsen** [comment]
@willfarrell ok well.... good news and bad news :)  Good news is that I added the tests with version pinning and a plugin. The bad news is that I can't reproduce the issue. I wonder if this is specific to Debian, rather than Debian-derivative distributions (Ubuntu).

**Comment by johnpneumann** [comment]
@leifmadsen - I was able to reproduce the issue with Ubuntu. The issue seems to be related to these 3 issues: [JENKINS-23232](https://issues.jenkins-ci.org/browse/JENKINS-23232), [JENKINS-23110](https://issues.jenkins-ci.org/browse/JENKINS-23110) and [JENKINS-35197](https://issues.jenkins-ci.org/browse/JENKINS-35197). 

The workaround for this would be to have something in `vars/Debian.yml` that specifies the `$JAVA_ARGS -Dhudson.diyChunking=false` as a sane default and then using combine to merge the `jenkins_init_changes` with it, but messing with someones environment seems bad to me. I solved it by overriding the `jenkins_init_changes` with:

```
jenkins_init_changes:
  - option: "JAVA_ARGS"
    value: "-Dhudson.diyChunking=false"
```

which worked for me.

**Comment by leifmadsen** [comment]
@johnpneumann awesome thanks for the info. I'm not entirely sure if there should be a change then... maybe just a documentation note? Should I actually attempt to address this?

CC: @geerlingguy

**Comment by geerlingguy** [comment]
@leifmadsen - I've merged your PR, but do you think should we do anything else prior to tagging a new minor release or will this be good enough for now, and we can take some time to make anything else better?

**Comment by leifmadsen** [comment]
@geerlingguy everything works for me, so any other issues are really somewhat outside the scope of this playbook directly (i.e. bugs in Jenkins or elsewhere).

I'm on CentOS, and clearly we want things to work for everyone, but we'd have to put in a bunch of work arounds, and I'm not sure that's really something I want to do (plus I don't use Ubuntu, so I can't really easily test).

From my viewpoint, nothing else is blocking a new release.

**Comment by geerlingguy** [comment]
@leifmadsen - Okay, thanks so much for your help! Going to tag 2.5.0 now.

**Comment by johnpneumann** [comment]
Would it be worth adding a note to the README referencing this issue for Debian/Ubuntu users?

**Comment by geerlingguy** [comment]
@johnpneumann - I'd rather try to find a good fix for it... can you open a separate issue that's Deb/Ubuntu specific and we can work from there?
